components/cqrs/event_processor.go

```diff
@@ -71,23 +149,76 @@ func NewEventProcessor(
 		logger = watermill.NopLogger{}
 	}
 
-	return &EventProcessor{
-		handlers,
-		generateTopic,
-		subscriberConstructor,
-		marshaler,
-		logger,
-	}, nil
+	eventProcessorConfig := EventProcessorConfig{
+		AckOnUnknownEvent: true, // this is the previous default behaviour - keeping backwards compatibility
+		GenerateSubscribeTopic: func(params EventProcessorGenerateSubscribeTopicParams) (string, error) {
+			return generateTopic(params.EventName), nil
+		},
+		SubscriberConstructor: func(params EventProcessorSubscriberConstructorParams) (message.Subscriber, error) {
+			return subscriberConstructor(params.HandlerName)
+		},
+		Marshaler: marshaler,
+		Logger:    logger,
+	}
+	eventProcessorConfig.setDefaults()
+
+	ep, err := NewEventProcessorWithConfig(eventProcessorConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, handler := range individualHandlers {
+		ep.AddHandlers(handler)
+	}
+
+	return ep, nil
+}
+
+// EventsSubscriberConstructor creates a subscriber for EventHandler.
+// It allows you to create separated customized Subscriber for every command handler.
+//
+// When handler groups are used, handler group is passed as handlerName.
+// Deprecated: please use EventProcessorSubscriberConstructorFn instead.
+type EventsSubscriberConstructor func(handlerName string) (message.Subscriber, error)
+
+// AddHandlers adds a new EventHandler to the EventProcessor.
+//
+// IMPORTANT: It's required to call AddHandlersToRouter to add the handlers to the router after calling AddHandlers.
+func (p *EventProcessor) AddHandlers(handler ...EventHandler) {
+	p.individualHandlers = append(p.individualHandlers, handler...)
 }
 
+// AddHandlersToRouter adds the EventProcessor's handlers to the given router.
+// It should be called only once per EventProcessor instance.
 func (p EventProcessor) AddHandlersToRouter(r *message.Router) error {
-	for i := range p.Handlers() {
-		handler := p.handlers[i]
+	handlers := p.Handlers()
+	if len(handlers) == 0 {
+		return errors.New("EventProcessor has no handlers, did you call AddHandlers?")
+	}
+
+	for i := range p.individualHandlers {
+		handler := p.individualHandlers[i]
+
+		if err := validateEvent(handler.NewEvent()); err != nil {
+			return errors.Wrapf(err, "invalid event for handler %s", handler.HandlerName())
+		}
+
 		handlerName := handler.HandlerName()
-		eventName := p.marshaler.Name(handler.NewEvent())
-		topicName := p.generateTopic(eventName)
+		eventName := p.config.Marshaler.Name(handler.NewEvent())
 
-		logger := p.logger.With(watermill.LogFields{
+		if p.config.GenerateSubscribeTopic == nil {
```

Comment:
Could be moved before the loop or deleted as the `validate` checks it.

Reply:
leftover, removed